process/block/preprocess/transactions.go

```diff
@@ -220,65 +232,249 @@ func (txs *transactions) ProcessBlockTransactions(
+
+	allTxs := make([]*txcache.WrappedTransaction, 0, process.MaxItemsInBlock)
+
+	for _, miniBlock := range block.MiniBlockSlice(body) {
+		shouldSkipMiniblock := miniBlock.SenderShardID != txs.shardCoordinator.SelfId() || !txs.isMiniBlockCorrect(miniBlock.Type)
+		if shouldSkipMiniblock {
 			continue
 		}
 
-		gasConsumedByMiniBlockInSenderShard := uint64(0)
-		gasConsumedByMiniBlockInReceiverShard := uint64(0)
+		txsFromMiniBlock, err := txs.computeTxsFromMiniBlock(miniBlock)
+		if err != nil {
+			return nil, err
+		}
 
-		for j := 0; j < len(miniBlock.TxHashes); j++ {
-			if !haveTime() {
-				return process.ErrTimeIsOut
-			}
+		allTxs = append(allTxs, txsFromMiniBlock...)
+	}
+
+	return allTxs, nil
+}
+
+func (txs *transactions) computeTxsFromMiniBlock(miniBlock *block.MiniBlock) ([]*txcache.WrappedTransaction, error) {
+	txsFromMiniBlock := make([]*txcache.WrappedTransaction, 0, process.MaxItemsInBlock)
+
+	for i := 0; i < len(miniBlock.TxHashes); i++ {
+		txHash := miniBlock.TxHashes[i]
+		txs.txsForCurrBlock.mutTxsForBlock.RLock()
+		txInfoFromMap := txs.txsForCurrBlock.txHashAndInfo[string(txHash)]
+		txs.txsForCurrBlock.mutTxsForBlock.RUnlock()
+
+		if txInfoFromMap == nil || txInfoFromMap.tx == nil {
+			log.Debug("missing transaction in computeTxsFromMiniBlock", "type", miniBlock.Type, "txHash", txHash)
+			return nil, process.ErrMissingTransaction
+		}
+
+		tx, ok := txInfoFromMap.tx.(*transaction.Transaction)
+		if !ok {
+			return nil, process.ErrWrongTypeAssertion
+		}
+
+		calculatedSenderShardId, err := txs.getShardFromAddress(tx.GetSndAddress())
+		if err != nil {
+			return nil, err
+		}
+
+		calculatedReceiverShardId, err := txs.getShardFromAddress(tx.GetRecvAddress())
+		if err != nil {
+			return nil, err
+		}
+
+		wrappedTx := &txcache.WrappedTransaction{
+			Tx:              tx,
+			TxHash:          txHash,
+			SenderShardID:   calculatedSenderShardId,
+			ReceiverShardID: calculatedReceiverShardId,
+		}
+
+		txsFromMiniBlock = append(txsFromMiniBlock, wrappedTx)
+	}
+
+	return txsFromMiniBlock, nil
+}
+
+func (txs *transactions) getShardFromAddress(address []byte) (uint32, error) {
+	isEmptyAddress := bytes.Equal(address, make([]byte, txs.addressConverter.AddressLen()))
+	if isEmptyAddress {
+		return txs.shardCoordinator.SelfId(), nil
+	}
+
+	addressContainer, err := txs.addressConverter.CreateAddressFromPublicKeyBytes(address)
+	if err != nil {
+		return 0, err
+	}
+
+	return txs.shardCoordinator.ComputeId(addressContainer), nil
+}
+
+func (txs *transactions) processTxsToMe(
+	body block.Body,
+	haveTime func() bool,
+) error {
+
+	startTime := time.Now()
+	defer func() {
+		elapsedTime := time.Since(startTime)
+		log.Debug("elapsed time to processTxsToMe",
+			"time [s]", elapsedTime,
+		)
+	}()
+
+	txsToMe, err := txs.computeTxsToMe(body)
+	if err != nil {
+		return err
+	}
+
+	//TODO: Should be verified if gas computation should be done here (before the next for) and should be done
```

Comment:
gas computation seems correct, should be left as it is.

Reply:
ok